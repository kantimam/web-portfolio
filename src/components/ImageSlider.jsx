import React, { Component } from 'react'


export default class ImageSlider extends Component {
    transitionInterval=null;
    activeImage=0;
    state={
        topImage: this.props.images[0],
        bottomImage: null,
        transition: 0
    }

    switchImage=()=>{
        const {images=[], duration=2000}=this.props
        if (images.length < 2) return;

        if(this.transitionInterval!==null) return

        const durationSeconds=duration / 1000;
        
        const bottom = images[this.activeImage + 1] ? images[this.activeImage + 1] : images[0];
        
        this.setState({bottomImage: bottom})

        this.transitionInterval=setInterval(()=>{

            this.setState({transition: this.state.transition+(1 / (60 * durationSeconds))});
        }, 1000 / 60)
    }
    
    componentDidUpdate(prevProps){
        if(prevProps.active!==this.props.active){
            if(this.props.active) return this.switchImage();
        }
        if(this.state.transition>=1){
            this.finishTransition();
            if(this.props.active && !this.transitionInterval) this.switchImage();
        }
        
    }


    componentWillUnmount(){
        this.stopTransition();
    }

    stopTransition=()=>this.transitionInterval && clearInterval(this.transitionInterval);

    finishTransition=()=>{
        this.stopTransition();
        this.transitionInterval=null;
        this.setState({topImage: this.state.bottomImage, bottomImage: null, transition: 0});
        this.activeImage=(this.activeImage+1) % this.props.images.length;
    }

    render() {

        return (
            <div className="imageSlider">
                <div className="imageSliderContainer">
                    {this.state.bottomImage && <img style={{ opacity: this.state.transition }} src={this.state.bottomImage} alt="bottomImage" />}
                    <img src={this.state.topImage} style={{ opacity: 1 - this.state.transition }} alt="topImage" />
                </div>
            </div>
        )
    }
}
