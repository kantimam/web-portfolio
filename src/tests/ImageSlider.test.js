import React from 'react';
import {
    unmountComponentAtNode,
    render
} from 'react-dom';
import {
    act
} from "react-dom/test-utils";
import ImageSlider from '../components/ImageSlider'
import { shallow } from 'enzyme'



let container = null;
beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

it("handle 0 or 1 image", () => {
    act(() => {
        //render(<ImageSlider />, container)
        expect(() => ImageSlider()).toThrow(TypeError);

    })

    const imageUrl = "https://i.imgur.com/TCMSH9Z.png"
    act(() => {
        render(<ImageSlider images={[imageUrl]} />, container)
    })

    expect(container).toMatchSnapshot();

    expect(container.querySelector("img").src).toBe(imageUrl);

    const imageTwo = "https://i.imgur.com/rtA9EKz.jpg"
    act(() => {
        render(<ImageSlider images={[imageTwo, imageUrl]} active={false} />, container)
    })

    expect(container).toMatchSnapshot();

    expect(container.querySelector("img").src).toBe(imageUrl);

    //expect(container.querySelector(".npmIconLink").getAttribute("href")).toBe("https://github.com")


})


it("handle multiple images", () => {

    const imageUrl = "https://i.imgur.com/TCMSH9Z.png"
    const imageTwo = "https://i.imgur.com/rtA9EKz.jpg"
    act(() => {
        render(<ImageSlider images={[imageTwo, imageUrl]} active={false} />, container)
    })

    expect(container).toMatchSnapshot();


    expect(container.querySelector("img").src).toBe(imageTwo);

})


it("check state when inactive", async () => {
    let wrapper;
    const imageUrl = "https://i.imgur.com/TCMSH9Z.png"
    const imageTwo = "https://i.imgur.com/rtA9EKz.jpg"
    act(() => {
        wrapper = shallow(<ImageSlider images={[imageTwo, imageUrl]} active={false} />)
    })

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.state("bottomImage")).not.toBeDefined();

    expect(wrapper.find("img").prop("src")).toEqual(imageTwo);

    expect(wrapper.state("topImage")).toEqual(imageTwo);

    expect(wrapper.state("bottomImage")).not.toBeDefined();

    await wrapper.setState({ bottomImage: "test" });
    expect(wrapper.state("bottomImage")).toEqual("test");

})

it("check state when active", async () => {
    let wrapper;
    const imageUrl = "https://i.imgur.com/TCMSH9Z.png"
    const imageTwo = "https://i.imgur.com/rtA9EKz.jpg"
    act(() => {
        wrapper = shallow(<ImageSlider duration={4000} images={[imageTwo, imageUrl]} active={true} />)
    })

    expect(wrapper).toMatchSnapshot();


    expect(wrapper.find("img").last().prop("src")).toEqual(imageTwo);

    expect(wrapper.state("topImage")).toEqual(imageTwo);

    expect(wrapper.state("bottomImage")).toBeDefined();

    await new Promise(res => setTimeout(() => {
        expect(wrapper.state("topImage")).toEqual(imageUrl);
        expect(wrapper.state("transition")).not.toEqual(0);
        res()

    }, 5000))

    await wrapper.setState({ bottomImage: "test" });
    expect(wrapper.state("bottomImage")).toEqual("test");

})