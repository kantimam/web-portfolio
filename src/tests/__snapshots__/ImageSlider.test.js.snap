// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check state when active 1`] = `
<div
  className="imageSlider"
>
  <div
    className="imageSliderContainer"
  >
    <img
      alt="bottomImage"
      src="https://i.imgur.com/TCMSH9Z.png"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
    <img
      alt="topImage"
      src="https://i.imgur.com/rtA9EKz.jpg"
      style={
        Object {
          "opacity": 1,
        }
      }
    />
  </div>
</div>
`;

exports[`check state when inactive 1`] = `
<div
  className="imageSlider"
>
  <div
    className="imageSliderContainer"
  >
    <img
      alt="topImage"
      src="https://i.imgur.com/rtA9EKz.jpg"
      style={
        Object {
          "opacity": 1,
        }
      }
    />
  </div>
</div>
`;

exports[`handle 0 or 1 image 1`] = `
<div>
  <div
    class="imageSlider"
  >
    <div
      class="imageSliderContainer"
    >
      <img
        alt="topImage"
        src="https://i.imgur.com/TCMSH9Z.png"
        style="opacity: 1;"
      />
    </div>
  </div>
</div>
`;

exports[`handle 0 or 1 image 2`] = `
<div>
  <div
    class="imageSlider"
  >
    <div
      class="imageSliderContainer"
    >
      <img
        alt="topImage"
        src="https://i.imgur.com/TCMSH9Z.png"
        style="opacity: 1;"
      />
    </div>
  </div>
</div>
`;

exports[`handle multiple images 1`] = `
<div>
  <div
    class="imageSlider"
  >
    <div
      class="imageSliderContainer"
    >
      <img
        alt="topImage"
        src="https://i.imgur.com/rtA9EKz.jpg"
        style="opacity: 1;"
      />
    </div>
  </div>
</div>
`;
